### PR TITLE
feat: refactor module resolution

### PR DESCRIPTION
**Pull request by diervo**
## Details
Make the module resolution explicit (via configuration) to improve performance and avoid ambiguity.

The format of `lwc.config.json` is as follows:
```json5
{
  modules: [
    "@lwc/engine", // can point to npm packages (must contain config in package.json)
    "src/modules", // can point to folders
    ["custom-module", "src/dist/custom/module.js"] //can point to custom files
  ]
}
```

## Does this PR introduce breaking changes?

* 🚨 `Yes, it does introduce breaking changes.`

The API of the rollup plugin is slightly different.

The old semantics still should be preserved in terms of default folder searching

### CODE REVIEW

File: packages/@lwc/module-resolver/src/index.ts [hunk 1]
```diff
@@ -6,154 +6,109 @@
...
-    if (!config) {
-        try {
-            config = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8')).lwc;
-        } catch (ignore) {
-            // ignore
+export function resolveModulesFromDir(modulesDir: string): RegistryEntry[] {
+    const namespaces = fs.readdirSync(modulesDir);
+    const resolvedModules: RegistryEntry[] = [];
+    namespaces.forEach(ns => {
+        // TODO: A better isDir() bellow
```
Comment (apapko) [inline packages/@lwc/module-resolver/src/index.ts]: should this be addressed this time?

File: packages/@lwc/module-resolver/src/index.ts [hunk 2]
```diff
@@ -6,154 +6,109 @@
...
-interface FlatEntry {
-    path: string;
+export interface ModuleResolverConfig {
+    rootDir: string;
+    modules: ModuleId[];
 }
 
-function createRegistryEntry(entry, moduleSpecifier, moduleName, moduleNamespace): RegistryEntry {
-    return { entry, moduleSpecifier, moduleName, moduleNamespace };
+export type ModuleId = string | [string, string];
```
Comment (apapko) [inline packages/@lwc/module-resolver/src/index.ts]: i would prefer splitting these into 2 distinct types to improve maintenance and readability

File: packages/@lwc/module-resolver/src/index.ts [hunk 3]
```diff
@@ -6,154 +6,109 @@
...
+export function resolveModulesFromDir(modulesDir: string): RegistryEntry[] {
+    const namespaces = fs.readdirSync(modulesDir);
+    const resolvedModules: RegistryEntry[] = [];
+    namespaces.forEach(ns => {
+        // TODO: A better isDir() bellow
+        if (!path.extname(ns)) {
+            const namespacedModuleDir = path.join(modulesDir, ns);
+            const modules = fs.readdirSync(namespacedModuleDir);
+            modules.forEach(moduleName => {
+                const moduleDir = path.join(namespacedModuleDir, moduleName);
```
Comment (apapko) [inline packages/@lwc/module-resolver/src/index.ts]: should this loop check for subdirectories to recursively resolve nested folders? For example, utils folder in the first level:
```js
dir/
    subdir
       subdir.js
       subdir.html
       utils/
            utils.js
```

File: packages/@lwc/rollup-plugin/README.md [hunk 4]
```diff
@@ -26,5 +24,5 @@ export default {
 
 -   `rootDir` (string, optional, default: `input`) - set the LWC module directory
 -   `sourcemap` (boolean, optional, default: `false`) - make the LWC compiler produce source maps
--   `resolveFromPackages` (boolean, optional, default: `true`) - let the rollup plugin resolve modules from the `node_modules` directory
+-   `modules` Mapping of module specififers.
```
Comment (apapko) [inline packages/@lwc/rollup-plugin/README.md]: typo: specifiers

File: packages/@lwc/module-resolver/src/index.ts (hunk 1 above)
Comment (diervo) [reply packages/@lwc/module-resolver/src/index.ts]: I was in the plane, I just couldn't figure out the APIs xD

File: packages/@lwc/module-resolver/src/index.ts (hunk 2 above)
Comment (diervo) [reply packages/@lwc/module-resolver/src/index.ts]: Fair enough

**Comment by caridy** [comment]
just to clarify, for `["custom-module", "src/dist/custom/module.js"]`, does it means that there is a module specifier `"custom-module"` that will point to file in `"src/dist/custom/module.js"`?

**Comment by diervo** [comment]
@caridy correct

File: packages/@lwc/module-resolver/src/index.ts (hunk 3 above)
Comment (diervo) [reply packages/@lwc/module-resolver/src/index.ts]: No a module must have the entry point at the root level

File: packages/@lwc/module-resolver/src/index.ts [hunk 5]
```diff
@@ -6,154 +6,113 @@
...
-    if (modulePaths) {
-        return modulePaths;
-    }
-    if (!moduleDirectories && !rootDir) {
-        // paths() is spec'd to return null only for built-in node
-        // modules like 'http'. To be safe, return empty array in
-        // instead of null in this case.
-        return require.resolve.paths('.') || [];
+        if (fs.existsSync(lwcConfigFile)) {
+            resolvedModules = resolveModules({ rootDir: lwcConfigFile });
```
Comment (pmdartus) [inline packages/@lwc/module-resolver/src/index.ts]: @diervo Can you confirm that the expected behavior is that if you have both a `lwc.config.js` and a `package.json` the config file takes precedence over the `package.json` config?

File: packages/@lwc/module-resolver/src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        modules: config.modules || [],
+        rootDir,
+    };
+}
+
+export function mergeModules(userModules: ModuleSpecifier[], configModules: ModuleSpecifier[]) {
+    return Array.from(new Set(userModules.concat(configModules)));
+}
+
+export const LWC_CONFIG_FILE = 'lwc.config.json';
```
Comment (pmdartus) [inline packages/@lwc/module-resolver/src/utils.ts]: Move the constant at the top of the module.

File: packages/@lwc/module-resolver/src/index.ts [hunk 7]
```diff
@@ -6,154 +6,113 @@
...
-        const { mappings, visited, moduleRoot } = opts;
-        if (typeof modules === 'string') {
-            const packageEntries = resolveModulesInDir(path.join(moduleRoot, modules));
-            Object.keys(packageEntries).forEach(moduleName => {
-                if (!hasModuleBeenVisited(moduleName, visited)) {
-                    mappings[moduleName] = packageEntries[moduleName];
-                    visited.add(moduleName);
-                }
-            });
+export function resolveModulesFromList(
```
Comment (pmdartus) [inline packages/@lwc/module-resolver/src/index.ts]: I don't think we should expose this API out of the package, the `resolveModules` API should be sufficient.

File: packages/@lwc/module-resolver/src/index.ts [hunk 8]
```diff
@@ -6,154 +6,113 @@
...
-
-            return mappings;
-        }, {});
+    return resolvedModules;
 }
 
-function hasModuleBeenVisited(module, visited) {
-    return visited.has(module);
-}
+export function resolveModulesFromNpm(packageName: string): RegistryEntry[] {
```
Comment (pmdartus) [inline packages/@lwc/module-resolver/src/index.ts]: If we were to do a major version bump, I think we should also remove this API. The `resolveModules` should be enough IMO.

File: packages/@lwc/module-resolver/src/__tests__/resolve.modules.spec.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import path from 'path';
+import { resolveModulesFromDir, resolveModules } from '../index';
+
+const FIXTURE_MODULE_ENTRIES = ['ns/cssEntry', 'ns/htmlEntry', 'ns/jsEntry'];
+
+describe('resolve modules', () => {
```
Comment (dbleakley) [inline packages/@lwc/module-resolver/src/__tests__/resolve.modules.spec.ts]: A couple of other tests that might be relevant -

Via the `resolveModules` API, should I be able to override the configuration?  (just as a note: tried this locally, it doesn't appear to pass)
```
    it('with configuration overrides resolving to custom modules', () => {
        const moduleDir = path.join(__dirname, 'fixtures/custom-resolution');
        const modules = resolveModules({ rootDir: moduleDir, modules: [["custom-module", "custom/module-alt.js"]] });
        const specifiers = modules.map(m => m.specifier);
        const entries = modules.map(m => m.entry);
        expect(specifiers).toStrictEqual(["custom-module"]);
        expect(entries).toStrictEqual([path.join(moduleDir, "custom/module-alt.js")]);
    });
```

Using the programmatic API to resolve one or more modules without a configuration file.
```
    it('from api configuration ', () => {
        const modules = resolveModules({ modules: ["@lwc/engine"] });
        const specifiers = modules.map(m => m.specifier);
        expect(specifiers).toStrictEqual(['lwc']);
    });
```

**Comment by rtm** [comment]
I presume you have taken a look at the spec for the `paths` property in `tsconfig.json`, which deals with the same sort of issue, and in my mind is a little bit more flexible and capable syntax-wise, especially as regards its use of wildcards, and multiple targets. For instance, if this feature is to be used to handle a deviant (by LWC standards) directory structure where I want to keep some components inside a sub-directory, I think I might want to do something like this:

```
{
  "*": ["lwc", "lwc/my-feature"]
}
```

And hopefully that will tell the resolver that `c/foobar` should be looked for in `lwc/my-feature/foobar`, *and* tell SFDX that components inside `my-feature` should be compiled/uploaded/deployed.

A feature like this is likely to be helping (or tormenting) users for years to come, so important to get it right.

**Comment by rtm** [comment]
> Make the module resolution implicit 

Don't you mean "explicit"?

File: packages/@lwc/module-resolver/src/__tests__/resolve.modules.spec.ts (hunk 9 above)
Comment (diervo) [reply packages/@lwc/module-resolver/src/__tests__/resolve.modules.spec.ts]: added

**Pull request merged**